### PR TITLE
use utilnet.IsIPv6(), drop most uses of config.IPv6Mode

### PR DESCRIPTION
**Pull request by danwinship**
In a dual-stack world, code needs to decide whether to do IPv4-ish stuff or IPv6-ish stuff based on what sort of addresses it's actually working with, not based on a global switch. So this mostly moves us to that. (There are a few remaining uses of `config.IPv6Mode` which will be dealt with later.)

As part of this, I also switched some code to use `IsIPv6()/IsIPv6String()/IsIPv6CIDR()/IsIPv6CIDRString()` from `k8s.io/utils/net` rather than confusingly fiddling with `ip.To4() == nil` / `!= nil`, etc.

oh, also, the localnet code was previously adding an iptables rule

    -t nat -A POSTROUTING -s 169.254.33.2/24 -j MASQUERADE

which is technically invalid; the address should be either `169.254.33.2/32` or `169.254.33.0/24`. iptables interprets it as the latter, but I'm pretty sure we meant the former and this patch changes it to that. (The change is mostly cosmetic, other than that previously we would have been NATting both 169.254.33.2 and 169.254.33.1, but there shouldn't have been any traffic from the latter anyway.)

### CODE REVIEW

**Comment by dcbw** [comment]
> which is technically invalid; the address should be either `169.254.33.2/32` or `169.254.33.0/24`. iptables interprets it as the latter, but I'm pretty sure we meant the former and this patch changes it to that. (The change is mostly cosmetic, other than that previously we would have been NATting both 169.254.33.2 and 169.254.33.1, but there shouldn't have been any traffic from the latter anyway.)

@danwinship I would have expected the former (eg `.2/32`) actually... but I guess either one works?

**Comment by danwinship** [comment]
> I would have expected the former (eg `.2/32`) actually

Yeah, that's what it does. Well, it doesn't actually include the `/32` part but that's implied

File: go-controller/pkg/node/gateway_shared_intf.go [hunk 1]
```diff
@@ -253,7 +253,7 @@ func addDefaultConntrackRules(nodeName, gwBridge, gwIntf string) error {
 	return nil
 }
 
-func initSharedGateway(nodeName string, subnet, gwNextHop, gwIntf string,
+func initSharedGateway(nodeName string, subnet string, gwNextHop, gwIntf string,
```
Comment (trozet) [inline go-controller/pkg/node/gateway_shared_intf.go]: why is this changed?

Comment (danwinship) [reply go-controller/pkg/node/gateway_shared_intf.go]: oh, I had probably changed it to take a `*net.IPNet` in an earlier version and then reverted it back, but incorrectly

File: go-controller/pkg/node/gateway_localnet.go [hunk 2]
```diff
@@ -179,45 +156,76 @@ func initLocalnetGateway(nodeName string, subnet string, wf *factory.WatchFactor
...
-		err = localnetNodePortWatcher(ipt, wf)
+		err = localnetNodePortWatcher(ipt, wf, gatewayIP)
 	}
 
 	return err
 }
 
-func localnetIptRules(svc *kapi.Service) []iptRule {
+// localnetIPTablesHelper gets an IPTablesHelper for IPv4 or IPv6 as appropriate
+func localnetIPTablesHelper(subnet string) (util.IPTablesHelper, error) {
```
Comment (alexanderConstantinescu) [inline go-controller/pkg/node/gateway_localnet.go:214]: @danwinship : what does this mean for us in a dual-stack world with services which can be both IPv6 and IPv4? I guess this is incorrect in such a case? We should not define a global helper, but instead instantiate one for each service as to ascertain the correct IP family and act accordingly (which is the way shared-gateway mode does it: https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/node/gateway_shared_intf_linux.go#L167).

I can take care of this in my ExternalIP PR, but I just need your confirmation that my understanding is correct.

Comment (alexanderConstantinescu) [reply go-controller/pkg/node/gateway_localnet.go:214]: Or be smarter than that and define a map of iptables-helpers for each family and just re-use that for every service we process.
